### PR TITLE
@grafana/data: Matchers and Transforms

### PR DESCRIPTION
**Pull request by ryantxu**
This PR proposes two new key types aimed to let us manipulate DataFrame[] results in the frontend.  The APIs are designed to be easily driven by settings saved in json (ie, panel/dashboard configs).  The initial APIs are not focused on using these features from javascript -- though that should come in a later issue.

**Matchers** 
fieldMatcher and frameMatcher.  This is various config options that let you say if something should apply to a field or DataFrame:
```
export type FieldMatcher = (field: Field) => boolean;
export type FrameMatcher = (frame: DataFrame) => boolean;
```

**Transformers** 
This represents a function that will manipulate DataFrame[]:
```
export type DataTransformer = (data: DataFrame[]) => DataFrame[];
```
Transformations will support standard features like 'rows to columns', 'time series to rows' etc that already exist in the Table panel.

this PR does not aim to build the UI to configure the functions, but is designed with that in mind.

### CODE REVIEW

**Comment by torkelo** [comment]
I like where this is going!

File: packages/grafana-data/src/utils/matchers/matchers.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
+import { Field, DataFrame } from '../../types/dataFrame';
+import { Registry, RegistryItemWithOptions } from '../registry';
+
+export type DataMatcher = (series: DataFrame, field?: Field) => boolean;
```
Comment (dprokop) [inline packages/grafana-data/src/utils/matchers/matchers.ts]: @ryan going through the code. Seems like I get the feeling what what we are about to achieve with this concept.

One thing thad bugs me here is what is your thinking behind pushing series as an argument of a matcher?

Comment (ryantxu) [reply packages/grafana-data/src/utils/matchers/matchers.ts]: Perhaps we should have a FieldMatcher and a FrameMatchet?

**Comment by ryantxu** [comment]
@dprokop -- since you have looked at this I have:
1. renamed `calc` to `reduce`
2. split `matchers` into `FieldMatcher` and `FrameMatcher` -- this causes a bunch of duplication, but I think the distinction is better than doing both in one pass
3. removed the super early UI component to pick a matcher -- lets do that in a different issue
4. added a super basic api to `dataTransformers` to give an idea on how we could more easily call this from javascript

I think this is almost ready to commit and we can work on the UI and maybe api in follow on issues

**Comment by dprokop** [comment]
@ryantxu thanks, will have a look soon!

**Comment by dprokop** [comment]
@ryantxu went through the entire thing and it feels like it's ok to be merged. Not sure I understand the need for :

> added a super basic api to dataTransformers to give an idea on how we could more easily call this from javascript

correctly. You want the transformers to be available for devs directly as well? I was thinking more of a reusable UI for that. But after second though, these are/going to be some basic operations on data, so maybe having access to those (i.e. for plugins) makes sense.

I think we gonna need sth similar for matchers registry as well, as otherwise we there will be need fo using matcher's ID when passing this to i.e. reduce transformer. But we can work on that when working on #18740

**Review by dprokop — APPROVED** [review]

**Review by torkelo — APPROVED** [review]
Looks good!

**Pull request merged**